helix-core/src/main/java/org/apache/helix/messaging/handling/HelixStateTransitionHandler.java

```diff
@@ -272,24 +236,31 @@ void postHandleMessage() {
     }
 
     try {
-      // Update the ZK current state of the node
-      PropertyKey key = _isTaskMessage && !_isTaskCurrentStatePathDisabled ? accessor.keyBuilder()
-          .taskCurrentState(instanceName, sessionId, resource,
-              bucketizer.getBucketName(partitionKey)) : accessor.keyBuilder()
-          .currentState(instanceName, sessionId, resource, bucketizer.getBucketName(partitionKey));
-      if (_message.getAttribute(Attributes.PARENT_MSG_ID) == null) {
-        // normal message
-        if (!accessor.updateProperty(key, _currentStateDelta)) {
-          throw new HelixException("Fails to persist current state back to ZK for resource "
-              + resource + " partition: " + _message.getPartitionName());
+        // Update the ZK current state of the node
+        if (!_stateModel.getCurrentState().equals(_currentStateDelta.getState(partitionKey))
+            && !_currentStateDelta.getState(partitionKey)
+            .equalsIgnoreCase(HelixDefinedState.DROPPED.toString())) {
```

Comment:
Question: Why DROPPED case needs to be handled differently?

Reply:
This check is to make sure that when updating `CurrentState` ZNode, `TaskRunner` haven't touch `CurrentState` ZNode yet. 
In line 185 in the changedHelixStateTransitionHandler (or line 222 in original version), `_stateModel` is not updated in DROPPED case. So the first condition is always fail for DROPPED case.